### PR TITLE
Projectkk2glider/issue 3096 table layout optonal

### PR DESCRIPTION
**Pull request by projectkk2glider**
I think the best option for now is to disable the new table layout until the solution for the slowness is found. 

For that I prepared this pull that adds a compile time option to use or not the new table layout. The change in time is not so significant on my system (cca 2.5 times faster) but should be almost the same as before I added table layout.

The code when using grid layout is not 100% the same as before, but should be pretty similar.

Use `cmake -D TABLE_LAYOUT=yes ` to go back to table layout (for testing).

Related to issue #3096

### CODE REVIEW

File: companion/src/helpers.h [hunk 1]
```diff
@@ -210,9 +213,51 @@ class TableLayout
...
-  QTableWidget * tableWidget;  
+#if defined(TABLE_LAYOUT)
+  QTableWidget * tableWidget; 
+#else
+  QGridLayout * gridWidget; 
+#endif
 };
 
+
+class Stopwatch
```
Comment (bsongis) [inline companion/src/helpers.h]: Perhaps a separate file? But I will merge to avoid further conflicts!

**Comment by kilrah** [comment]
Hmm it looks much better for some reason... but there's also a lot of difference between release and debug builds, and whether I attach the debugger (needed to see the output in VC) or run normally.

Debug build, with attached debugger:

With `cmake -D TABLE_LAYOUT=yes `
```
"global" "ModelEdit creation" "0 ms [0 ms]" 
"global" "ModelEdit start constructor" "1 ms [1 ms]" 
"ModelEdit" "inputs" "811 ms [811 ms]" 
"ModelEdit" "Mixes" "3 ms [814 ms]" 
"Channels" "header" "46 ms [46 ms]" 
"Channels" "add elements" "1491 ms [1537 ms]" 
"Channels" "end" "20 ms [1557 ms]" 
"ModelEdit" "Outputs" "1560 ms [2374 ms]" 
"LogicalSwitchesPanel" "header" "10 ms [10 ms]" 
"LogicalSwitchesPanel" "added elements" "816 ms [826 ms]" 
"LogicalSwitchesPanel" "end" "4487 ms [5313 ms]" 
"ModelEdit" "LS" "5356 ms [7730 ms]" 
"CustomFunctionsPanel - populate" "get tracks" "1 ms [1 ms]" 
"CustomFunctionsPanel - populate" "get scripts" "0 ms [1 ms]" 
"CustomFunctionsPanel - populate" "add items" "5387 ms [5388 ms]" 
"CustomFunctionsPanel - populate" "disableMouseScrolling" "5 ms [5393 ms]" 
"CustomFunctionsPanel - populate" "update" "23 ms [5416 ms]" 
"CustomFunctionsPanel - populate" "resizeColumnsToContents" "23 ms [5439 ms]" 
"CustomFunctionsPanel - populate" "end" "0 ms [5439 ms]" 
"ModelEdit" "CF" "5447 ms [13177 ms]" 
"ModelEdit" "end" "764 ms [13941 ms]" 
"global" "ModelEdit end constructor" "13943 ms [13944 ms]" 
"global" "ModelEdit created" "0 ms [13944 ms]" 
"global" "STARTING MODEL EDIT" "0 ms [13944 ms]" 
"global" "ModelEdit shown" "77 ms [14021 ms]" 
```

With `cmake -D TABLE_LAYOUT=no `
```
"global" "ModelEdit start constructor" "0 ms [0 ms]" 
"ModelEdit" "inputs" "802 ms [802 ms]" 
"ModelEdit" "Mixes" "3 ms [805 ms]" 
"Channels" "header" "0 ms [0 ms]" 
"Channels" "add elements" "137 ms [137 ms]" 
"Channels" "end" "1 ms [138 ms]" 
"ModelEdit" "Outputs" "271 ms [1076 ms]" 
"LogicalSwitchesPanel" "header" "0 ms [0 ms]" 
"LogicalSwitchesPanel" "added elements" "63 ms [63 ms]" 
"LogicalSwitchesPanel" "end" "110 ms [173 ms]" 
"ModelEdit" "LS" "764 ms [1840 ms]" 
"CustomFunctionsPanel - populate" "get tracks" "1 ms [1 ms]" 
"CustomFunctionsPanel - populate" "get scripts" "0 ms [1 ms]" 
"CustomFunctionsPanel - populate" "add items" "519 ms [520 ms]" 
"CustomFunctionsPanel - populate" "disableMouseScrolling" "4 ms [524 ms]" 
"CustomFunctionsPanel - populate" "update" "4 ms [528 ms]" 
"CustomFunctionsPanel - populate" "resizeColumnsToContents" "0 ms [528 ms]" 
"CustomFunctionsPanel - populate" "end" "0 ms [528 ms]" 
"ModelEdit" "CF" "1322 ms [3162 ms]" 
"ModelEdit" "end" "452 ms [3614 ms]" 
"global" "ModelEdit end constructor" "3617 ms [3617 ms]" 
"global" "ModelEdit created" "0 ms [3617 ms]" 
"global" "STARTING MODEL EDIT" "0 ms [3617 ms]" 
"global" "ModelEdit shown" "47 ms [3664 ms]" 
```

Release build with attached debugger:

With `cmake -D TABLE_LAYOUT=yes `
```
"global" "ModelEdit creation" "0 ms [0 ms]" 
"global" "ModelEdit start constructor" "0 ms [0 ms]" 
"ModelEdit" "inputs" "255 ms [255 ms]" 
"ModelEdit" "Mixes" "4 ms [259 ms]" 
"Channels" "header" "27 ms [27 ms]" 
"Channels" "add elements" "508 ms [535 ms]" 
"Channels" "end" "6 ms [541 ms]" 
"ModelEdit" "Outputs" "543 ms [802 ms]" 
"LogicalSwitchesPanel" "header" "3 ms [3 ms]" 
"LogicalSwitchesPanel" "added elements" "290 ms [293 ms]" 
"LogicalSwitchesPanel" "end" "943 ms [1236 ms]" 
"ModelEdit" "LS" "1254 ms [2056 ms]" 
"CustomFunctionsPanel - populate" "get tracks" "0 ms [0 ms]" 
"CustomFunctionsPanel - populate" "get scripts" "0 ms [0 ms]" 
"CustomFunctionsPanel - populate" "add items" "1500 ms [1500 ms]" 
"CustomFunctionsPanel - populate" "disableMouseScrolling" "2 ms [1502 ms]" 
"CustomFunctionsPanel - populate" "update" "10 ms [1512 ms]" 
"CustomFunctionsPanel - populate" "resizeColumnsToContents" "8 ms [1520 ms]" 
"CustomFunctionsPanel - populate" "end" "0 ms [1520 ms]" 
"ModelEdit" "CF" "1525 ms [3581 ms]" 
"ModelEdit" "end" "223 ms [3804 ms]" 
"global" "ModelEdit end constructor" "3808 ms [3808 ms]" 
"global" "ModelEdit created" "0 ms [3808 ms]" 
"global" "STARTING MODEL EDIT" "0 ms [3808 ms]" 
"global" "ModelEdit shown" "26 ms [3834 ms]" 
```

With `cmake -D TABLE_LAYOUT=no `
```
"global" "ModelEdit start constructor" "0 ms [0 ms]" 
"ModelEdit" "inputs" "282 ms [282 ms]" 
"ModelEdit" "Mixes" "4 ms [286 ms]" 
"Channels" "header" "0 ms [0 ms]" 
"Channels" "add elements" "61 ms [61 ms]" 
"Channels" "end" "1 ms [62 ms]" 
"ModelEdit" "Outputs" "138 ms [424 ms]" 
"LogicalSwitchesPanel" "header" "0 ms [0 ms]" 
"LogicalSwitchesPanel" "added elements" "28 ms [28 ms]" 
"LogicalSwitchesPanel" "end" "48 ms [76 ms]" 
"ModelEdit" "LS" "404 ms [828 ms]" 
"CustomFunctionsPanel - populate" "get tracks" "0 ms [0 ms]" 
"CustomFunctionsPanel - populate" "get scripts" "0 ms [0 ms]" 
"CustomFunctionsPanel - populate" "add items" "223 ms [223 ms]" 
"CustomFunctionsPanel - populate" "disableMouseScrolling" "2 ms [225 ms]" 
"CustomFunctionsPanel - populate" "update" "1 ms [226 ms]" 
"CustomFunctionsPanel - populate" "resizeColumnsToContents" "0 ms [226 ms]" 
"CustomFunctionsPanel - populate" "end" "0 ms [226 ms]" 
"ModelEdit" "CF" "598 ms [1426 ms]" 
"ModelEdit" "end" "185 ms [1611 ms]" 
"global" "ModelEdit end constructor" "1615 ms [1615 ms]" 
"global" "ModelEdit created" "0 ms [1615 ms]" 
"global" "STARTING MODEL EDIT" "0 ms [1615 ms]" 
"global" "ModelEdit shown" "25 ms [1640 ms]" 
```

Release build with no attached debugger (stopwatch measurement), approx vs 2.5sec vs 1.2sec

So firstly the debugger has some influence, but mostly the build type does a lot.

**Comment by projectkk2glider** [comment]
All my times are posted with DEBUG build and not using debugger. I will try the release build and post my times.

**Comment by projectkk2glider** [comment]
Release build with grid layout:
```

**Comment by projectkk2glider** [comment]
Some interesting warnings when doing true Relese build :smile: 
```
/src/gui/Taranis/menu_general_calib.cpp:162:115: warning: array subscript is above array bounds [-Warray-bounds]
                   SWAP(reusableBuffer.calib.xpotsCalib[idx].steps[j], reusableBuffer.calib.xpotsCalib[idx].steps[k]);
```

**Comment by kilrah** [comment]
For what target?

**Comment by kilrah** [comment]
Yes, but which Taranis... as the number of pots is different between versions it might not happen with all :)

**Comment by projectkk2glider** [comment]
The warning output comes form the Companion build when compiled like so:
```
$ cmake -D CMAKE_BUILD_TYPE=Release ../src/
$ make
```

But I can't reproduce it on an older machine now (using gcc 4.6.3).

**Comment by projectkk2glider** [comment]
Error happens when building libopentx-taranisplus-simulator.so and libopentx-taranis-simulator.so. X9E is ok. (gcc 4.8.4)

**Comment by bsongis** [comment]
I don't see any bug there!
The funny thing is that the warning disappears if the first value of k in the loop is j. How is it possible?

**Comment by projectkk2glider** [comment]
Maybe it is a gcc bug, it happened before (http://eigen.tuxfamily.org/bz/show_bug.cgi?id=149, https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45180) and one is currently still open https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59124

**Comment by kilrah** [comment]
VC++ doesn't give a warning.

**Pull request merged**